rfcs/0052-create-sink-into-table.md

```diff
@@ -0,0 +1,399 @@
+
+```sql
+CREATE TABLE table_name AS select_query
+```
+
+### alter mv to table
+
+To make it more usable for users, I would like to introduce another function to alter an exiting materialzied view to a table.
+
+```sql
+ALTER MATERIALIZED VIEW mv_table TO TABLE
+```
+
+This is because now we still recommend `create materialized view` at the first place instead of `create table as`. However, application developers are facing changes all the time, so it's very likely that they may realize they need to 'alter' an existing materialized view. Meanwhile, the strong consistency gurantee of MV prevents us to do any altering. With the help of `ALTER MATERIALIZED VIEW TO TABLE`, users can choose to break this consistency gurantee and alter their streaming pipelines.
+
+Currently, we can just support it with limitation that the mv's primary key can not be hidden column, we will discuss why in the "implementation" phrase.
+
+The statement requires **Allow a sink and a table have the same name**. It is ok in our system because the sink will never appear in the FROM clause. An alternative is that we can add a `_sink` suffix on the sink's name.
+
+### Implementation
+
+Thanks for @BugenZhao and @shanicky's offline discussion.
+#### CREATE SINK into TABLE
+
+Currently, the table's plan is shown as the graph. In brief, the table's data can comes from the external source or BatchDML. And the DMLExec is used here as a side input. ![](./images/0052-sink-to-table/cur-table-plan.png).
+
+The `create sink to table` introduces more data sources for the table. And now RW has implemented stream Union executor and RowIdGen with specified distribution. So we can support it more clearly with these new features.
+![](./images/0052-sink-to-table/new-table-plan.png)
+
+Here are some Q&A for the details of the graph
+- What's the difference between "append only table" and "table without pk"?
+  - The only difference is that if Update/delte on the table is allowed. which make the non-append only table without pk
+    - not append only, so the watermark is not allowed on it.
+    - the DML changes must be hash shuffled to the materialize fragment.
+    - the materialize must handle pk conflict for the DML changes.
+  - maybe we will support alter table between append only and non-append only in future.  
+- Why do we need blackhole SinkExecutor here? 
+  - we have some special logic in the sink executor such as `force-append-only` and special compaction when the down stream's pk does not contain stream key, we can reuse these logic in SinkExecutor to maintain the same behvaior with other kind of sinks.
```

Comment:
I think this Executor is far different from a `SinkExecutor`. `SinkExecutor` has no output channel, and focused on executor the Sink of a particular external connector.

Actually, I think `force-append-only` is too simple and is not worth of reusing. And if you do want to reuse it, perhaps we can try to extract these logics into a new executor. Either way, I would definitely -1 for reusing `SinkExecutor`.

Reply:
In fact, the sink executor has output and can give downstream a processed stream. I guess we will have more and more common logic in the sink executor such as https://risingwave-labs.slack.com/archives/C048771BGKF/p1693553432021789.
 So. I think a blackhole sink here makes sense.

Reply:
Hmm, I remember that the reason why "the sink executor has output" is because that makes sink executors **non-special** - it's driven by the actor's `for` loop just like any other executors.

It's like using an undefined behavior, that's why I feel not so good.

Reply:
Emmm, so how about requiring the sink executor's output should be the data that the sink should emit? What logic I want to reuse is "parse the sink catalog and do some process on the input stream". I guess this should be done in the sink executor.

Reply:
> "parse the sink catalog and do some process on the input stream"

👍 It makes more sense to me.